File: Python_Scripting_for_Computational_Science_Third_Edition/py/regex/swap2.py
#!/usr/bin/env python
import re, glob, string
# as swap1.py except that we here use a regex with comments and compile it:
arg = r'[^,]+'
call = re.compile(r"""
         superLibFunc  # name of function to match 
         \s*           # optional whitespace
         \(            # parenthesis before argument list
         \s*           # optional whitespace
         (?P<arg1>%s)  # first argument plus optional whitespace
         ,             # comma between the arguments
         \s*           # optional whitespace
         (?P<arg2>%s)  # second argument plus optional whitespace
         \)            # closing parenthesis
         """ % (arg,arg), re.VERBOSE)

cfiles = ['.test1.c']
for cfile in cfiles:
    print 'Treating',cfile
    file = open(cfile, 'r')
    filestr = file.read() # load all lines into a string
    file.close()
    filestr = call.sub(r'superLibFunc(\g<arg2>, \g<arg1>)', filestr)
    file = open(cfile + '.tmp', 'w')
    file.write(filestr)  # print everything into cfile.tmp

File: Python_Scripting_for_Computational_Science_Third_Edition/perl/swap2.pl
#!/usr/bin/perl
# as swap1.pl but with comments inside regular expressions

$arg  = "[^,]+";
$call = "superLibFunc  # name of function to match 
           \\s*          # possible whitespace
           \\(           # left parenthesis
           \\s*          # possible whitespace
           ($arg)        # first argument plus optional whitespace
           ,             # comma between the arguments
           \\s*          # possible whitespace
           ($arg)        # second argument plus optional whitespace
           \\)           # closing parenthesis
           ";

@cfiles = ("../py/regex/.test1.c");
foreach $cfile (@cfiles) {
    print "Treating $cfile...\n";
    open(FILE,"<$cfile") || die "Problems: $!\n";
    @lines = <FILE>;  # load all lines into a list
    close(FILE);
    $filestr = join("",@lines);
    $filestr =~ s/$call/superLibFunc($2, $1)/gx;
    # can also write this as
    # $filestr =~ s/$call/superLibFunc(\2, \1)/gx;
    open(FILE,">$cfile.tmp");
    print FILE $filestr;

    # write out the result of a match (to demonstrate use of $1, $2, and $&):
    foreach $line (@lines) {
        if ($line =~ /$call/xg) {
            print "line: $line    match=$&  arg1=$1  arg2=$2\n";
	}
    }
}
